File: DotNetGotchas/VBDN/ParamThreadSafety/Modified4/ThreadPassingParams/ALib/SomeClass.vb
'SomeClass.vb part of ALib.dll
Imports System.Threading

Public Class SomeClass

    Private Sub Method1(ByVal val As Integer)
        Console.WriteLine( _
            "Method1 runs on Thread {0} with {1}", _
            AppDomain.GetCurrentThreadId(), val)

        ' Some operation takes place here
    End Sub

    Private Delegate Sub CallMethod1Delegate(ByVal val As Integer)


    Public Sub DoSomething(ByVal val As Integer)
        ' Some operation...

        ' Want to call Method1 in different thread
        ' from here?
        Dim theRunner As New ThreadRunner( _
            New CallMethod1Delegate(AddressOf Method1), val)

        theRunner.Start()

        ' Some operation...
    End Sub
End Class

File: DotNetGotchas/CSharp/ParamThreadSafety/Modified4/ThreadPassingParams/ALib/SomeClass.cs
//SomeClass.cs part of ALib.dll
using System;
using System.Threading;

namespace ALib
{
	public class SomeClass
	{
		private void Method1(int val)
		{
			// Some operation takes place here
			Console.WriteLine(
				"Method1 runs on Thread {0} with {1}",
				AppDomain.GetCurrentThreadId(), val);
		}

		private delegate void CallMethod1Delegate(int val);

		public void DoSomething(int val)
		{
			// Some operation...

			// Want to call Method1 in different thread
			// from here?
			ThreadRunner theRunner 
				= new ThreadRunner(
					new CallMethod1Delegate(Method1), val);

			theRunner.Start();

			// Some operation...
		}
	}
}
